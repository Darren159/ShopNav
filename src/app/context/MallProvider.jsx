import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import PropTypes from "prop-types";
import fetchMalls from "../../services/fetchMalls";

export const MallContext = createContext();

export function MallProvider({ children }) {
  const [malls, setMalls] = useState([]);
  const [currentMall, setCurrentMall] = useState(null);

  useEffect(() => {
    const initializeMalls = async () => {
      try {
        const mallList = await fetchMalls();
        setMalls(mallList);
        setCurrentMall(mallList[0]); // Set the first mall as the current mall
      } catch (error) {
        Alert.alert("Error", error.message, [{ text: "OK" }], {
          cancelable: false,
        });
      }
    };
    initializeMalls();
  }, []);
  const value = useMemo(
    () => ({ malls, currentMall, setCurrentMall }),
    [malls, currentMall]
  );

  return <MallContext.Provider value={value}>{children}</MallContext.Provider>;
}

MallProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
