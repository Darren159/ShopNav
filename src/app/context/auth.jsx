import { createContext, useState, useEffect, useMemo } from "react";
import { Alert } from "react-native";
import {
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import PropTypes from "prop-types";

import { auth } from "../../../firebaseConfig";

export const AuthContext = createContext({});

export function AuthProvider({ children }) {
  const [user, setUser] = useState(null);

  useEffect(() => {
    onAuthStateChanged(auth, (userState) => {
      if (userState) {
        setUser(userState);
      } else {
        setUser(null);
      }
    });
  }, []);

  const authContextValue = useMemo(
    () => ({
      user,
      setUser,
      signin: async (email, password) => {
        try {
          await signInWithEmailAndPassword(auth, email, password);
        } catch (error) {
          let errorMessage;
          
          switch (error.code) {
            case 'auth/user-not-found':
              errorMessage = 'There is no user record corresponding to this identifier. The user may have been deleted.';
              break;
            case 'auth/wrong-password':
              errorMessage = 'The password is invalid or the user does not have a password.';
              break;
            case 'auth/invalid-email':
              errorMessage = 'The email address is badly formatted.';
              break;
            default:
              errorMessage = 'Something went wrong. Please try again later.';
          }
      
          Alert.alert(
            "Authentication Failed",
            errorMessage,
            [
              { text: "OK" }
            ]
          );
        }
      },
      signout: async () => {
        try {
          await signOut(auth);
        } catch (e) {
          console.error(e);
        }
      },
    }),
    [user]
  );

  return (
    <AuthContext.Provider value={authContextValue}>
      {children}
    </AuthContext.Provider>
  );
}

AuthProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
