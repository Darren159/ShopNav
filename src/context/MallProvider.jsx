import { createContext, useState, useEffect, useMemo } from "react";
import PropTypes from "prop-types";
import { collection, getDocs } from "firebase/firestore";
import { db } from "../../firebaseConfig";

export const MallContext = createContext();

export function MallProvider({ children }) {
  const [malls, setMalls] = useState([]);
  const [currentMall, setCurrentMall] = useState(null);

  useEffect(() => {
    const fetchMalls = async () => {
      const mallCollection = collection(db, "malls");
      const mallSnapshot = await getDocs(mallCollection);
      const mallList = mallSnapshot.docs.map((mall) => mall.id);
      setMalls(mallList);
      setCurrentMall(mallList[0]); // Set the first mall as the current mall
    };
    fetchMalls();
  }, []);
  const value = useMemo(
    () => ({ malls, currentMall, setCurrentMall }),
    [malls, currentMall]
  );

  return <MallContext.Provider value={value}>{children}</MallContext.Provider>;
}

MallProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
