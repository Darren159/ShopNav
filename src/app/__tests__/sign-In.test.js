import renderer from "react-test-renderer";
import { Alert } from "react-native";
import { router } from "expo-router";
import { render, waitFor, fireEvent } from "@testing-library/react-native";
import SignIn from "../sign-in";
import { AuthContext } from "../context/auth";

jest.mock("firebase/auth", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../../../firebaseConfig.js", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("expo-router", () => ({
  Stack: {
    Screen: () => "MockedScreenComponent",
  },
  router: {
    replace: jest.fn(),
  },
}));

jest.spyOn(Alert, "alert");

const signInMock = jest.fn();

describe("SignIn", () => {
  it("renders correctly", async () => {
    const tree = renderer
      .create(
        <AuthContext.Provider value={{ user: null, signin: signInMock }}>
          <SignIn />
        </AuthContext.Provider>
      )
      .toJSON();
    await waitFor(() => expect(tree).toMatchSnapshot());
  });

  it("calls signin function with the input value", async () => {
    const { getByText, getByPlaceholderText } = render(
      <AuthContext.Provider value={{ user: null, signin: signInMock }}>
        <SignIn />
      </AuthContext.Provider>
    );

    fireEvent.changeText(getByPlaceholderText("Email"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "password");
    fireEvent.press(getByText("Login"));

    await waitFor(() => {
      expect(signInMock).toHaveBeenCalledWith("test@example.com", "password");
    });
  });

  it("calls router function when there is a user logged in", () => {
    render(
      <AuthContext.Provider value={{ user: true, signin: signInMock }}>
        <SignIn />
      </AuthContext.Provider>
    );

    expect(router.replace).toHaveBeenCalledWith("/developerAccess");
  });

  it("shows an error alert when signin fails", async () => {
    const { getByText, getByPlaceholderText } = render(
      <AuthContext.Provider
        value={{
          user: null,
          signin: jest.fn(() => {
            throw new Error("Test error");
          }),
        }}
      >
        <SignIn />
      </AuthContext.Provider>
    );

    fireEvent.changeText(getByPlaceholderText("Email"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "password");
    fireEvent.press(getByText("Login"));

    await waitFor(() =>
      expect(Alert.alert).toHaveBeenCalledWith(
        "Sign In Error",
        "Invalid Email/Password",
        [{ text: "OK" }],
        { cancelable: false }
      )
    );
  });
  it("shows loading indicator when signing in", async () => {
    const { getByText, getByPlaceholderText, queryByTestId } = render(
      <AuthContext.Provider value={{ user: null, signin: signInMock }}>
        <SignIn />
      </AuthContext.Provider>
    );

    fireEvent.changeText(getByPlaceholderText("Email"), "test@example.com");
    fireEvent.changeText(getByPlaceholderText("Password"), "password");
    fireEvent.press(getByText("Login"));

    expect(queryByTestId("loading")).toBeTruthy();

    await waitFor(() => expect(queryByTestId("loading")).toBeNull());
  });
});
