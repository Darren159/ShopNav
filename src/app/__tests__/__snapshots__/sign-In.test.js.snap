// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn renders correctly 1`] = `
[
  "MockedScreenComponent",
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCompleteType="email"
      keyboardType="email-address"
      onChangeText={[Function]}
      placeholder="Email"
      style={
        {
          "borderColor": "grey",
          "borderRadius": 10,
          "borderWidth": 1,
          "marginBottom": 5,
          "padding": 10,
          "textAlign": "left",
          "width": "80%",
        }
      }
      textContentType="emailAddress"
      value=""
    />
    <TextInput
      autoCapitalize="none"
      autoCompleteType="password"
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        {
          "borderColor": "grey",
          "borderRadius": 10,
          "borderWidth": 1,
          "marginBottom": 5,
          "padding": 10,
          "textAlign": "left",
          "width": "80%",
        }
      }
      textContentType="password"
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#8AB9EF",
          "borderColor": "grey",
          "borderRadius": 10,
          "marginTop": 10,
          "opacity": 1,
          "padding": 10,
          "width": "80%",
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>,
]
`;
